✨ add `pandas.Series`

EkberHasanov opened the pull request:
* add pd.Series support to the pydantic_extra_types.pandas_types module
* created tests for series type
* add docs for basic usage


Selected Reviewer: @Kludex

EkberHasanov committed:
Add support and test for pd.Series

EkberHasanov committed:
Adding docs for basic usage

EkberHasanov committed:
Fix documentation

EkberHasanov committed:
Add pandas to optional dependencies

EkberHasanov committed:
fix python3.8 issues

EkberHasanov committed:
fix py3.7 issues

EkberHasanov committed:
fix macos error

EkberHasanov committed:
fix indentation error in ci.yml

yezz123 commented:
Thanks for the initiative @EkberHasanov its look a good PR and worth to work in, as i can see the issue is regarding py37 and missing dependencies do you think its worth re-open it and work on it ?

EkberHasanov commented:
Thank you for your feedback, @yezz123 . I appreciate your positive remarks about the PR. I'm definitely interested in working on it further and ensuring that it works flawlessly without any remaining issues, but I don't know should I continue to change something and push it, because it failed so much and frustrated me.

EkberHasanov committed:
ci: fix dependency issue in macos

EkberHasanov committed:
improve test coverage to 100%

yezz123 committed:
🔧 update code

yezz123 committed:
🍱 update requirements

yezz123 committed:
🍱 Fix Requirements

yezz123 committed:
🍱 Fix Requirements

yezz123 commented:
please review

EkberHasanov commented on pydantic_extra_types/pandas_types.py:
```diff
@@ -0,0 +1,49 @@
+    import pandas as pd
+except ModuleNotFoundError:  # pragma: no cover
+    raise RuntimeError(
+        '`PhoneNumber` requires "phonenumbers" to be installed. You can install it with "pip install phonenumbers"'
```
```suggestion
        '`Series` requires "pandas" to be installed. You can install it with "pip install pandas"'
```

yezz123 committed:
Update pydantic_extra_types/pandas_types.py

Co-authored-by: Akbar <98239031+EkberHasanov@users.noreply.github.com>

Kludex commented on pyproject.toml:
```diff
@@ -39,14 +39,15 @@ classifiers = [
 requires-python = '>=3.7'
 dependencies = [
-    'pydantic>=2.0b3',
+    'pydantic>=2.0',
```
Please open a new PR for this.
```suggestion
    'pydantic>=2.0b3',
```

Kludex commented on pyproject.toml:
```diff
@@ -39,14 +39,15 @@ classifiers = [
 all = [
     'phonenumbers>=8,<9',
     'pycountry>=22,<23',
+    'pandas==1.3.5',
```
Why not `pandas>=2.0`?

EkberHasanov replied:
Because `pandas>=2.0` requires python >=3.8

Kludex commented on pydantic_extra_types/pandas_types.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class Series:
+    def __init__(self, value: Any) -> None:
```
What is this "value"? `pd.Series` accepts many arguments. 🤔

Kludex commented on pydantic_extra_types/pandas_types.py:
```diff
@@ -0,0 +1,49 @@
+    ) -> core_schema.AfterValidatorFunctionSchema:
+        return core_schema.general_after_validator_function(
+            cls._validate,
+            core_schema.any_schema(),
```
I don't think this is the best we can do. Is it? A series should be an `array` of some type, and I guess we are able to get this type as well.

Kludex commented on requirements/linting.in:
```diff
@@ -4,3 +4,4 @@ annotated-types
 black
 pyupgrade
 ruff
+pandas-stubs
```
Likely not needed on pandas >= 2.

Kludex reviewed (changes requested):
I'm unsure if we should make it possible to instantiate `Series` with `__init__`.

Maybe it's better to not have the `__init__`, and only implement the `__get_pydantic_core_schema__`.

EkberHasanov commented:
> I'm unsure if we should make it possible to instantiate `Series` with `__init__`.
> 
> Maybe it's better to not have the `__init__`, and only implement the `__get_pydantic_core_schema__`.

Oh yes, I guess inheriting the `Series` class from `pd.Series` makes this implementation much cleaner and more aligned with the existing functionality of `pd.Series`. Regarding the `__init__` method, you're absolutely right that it may not be necessary. Since we are inheriting from `pd.Series`, we can rely on its default behavior for instantiation.

Thank you for pointing out this alternative approach. I'm excited about this implementation as it provides a more elegant and intuitive way to work with series data using Pydantic. Thank you for your feedback!

EkberHasanov committed:
Inheriting directly from pd.Series

EkberHasanov committed:
delete extra files

dzmitry-lahoda commented:
@EkberHasanov how it is going? should somebody catch up and finish this PR? thank you

EkberHasanov commented:
> @EkberHasanov how it is going? should somebody catch up and finish this PR? thank you

Hi @dzmitry-lahoda it's all good, thanks! I have made some changes last time, but i guess i have to update library versions now. I will give it a try to update the versions of the libraries and see what is going on, and then i will let you you know if I need further help. Thanks

EkberHasanov committed:
upgrading version of pandas

EkberHasanov committed:
Update test_json_schema.py

EkberHasanov committed:
fixing some issues

EkberHasanov committed:
change core_schema func

EkberHasanov committed:
Update test_json_schema.py

EkberHasanov commented:
Hi @dzmitry-lahoda, hope you're okay. I guess, I may need help to resolve all these issues. Could someone please help or perhaps explain what and where is wrong? I though I know, but it seems i'm wrong.

Commit:
Restore non-pandas files to match upstream/main

Commit:
build: add pandas optional dependency group

Commit:
feat: add Series pydantic type with generic validation

Implements Series[T] with pydantic-core no_info_wrap_validator_function,
adds stubs for Index and DataFrame, and replaces test file with full
test suite covering untyped/typed Series, Index, and DataFrame scenarios.

Commit:
feat: add Index pydantic type with generic validation

Commit:
feat: add DataFrame pydantic type with TypedDict column validation

Commit:
docs: update pandas_types docs for Series, Index, DataFrame

Commit:
style: apply ruff format to pandas_types

Commit:
build: regenerate uv.lock with pandas dependency

Commit:
fix: add type: ignore[import-untyped] for pandas, remove unused override ignore

EkberHasanov commented:
@yezz123 @Kludex @dzmitry-lahoda  - wassup gents, I guess I am finally done with this PR :). I decided to put more hours into open-source, so decided to get this unfinished "business" done and then move other projects and issues
